_posts/phoenix-reverse-proxy/2022-02-17-phoenix-reverse-proxy.md

```diff
@@ -0,0 +1,425 @@
+
+
+---
+layout: post
+title:  "A look into routing traffic into your umbrella and a deep-dive into the PhoenixReverseProxy"
+date:   2022-02-17 01:23:45 +0500
+categories: reverse proxy umbrella
+---
+
+# Problem
+When Phoenix projects grow, an umbrella structure is needed to separate different services. Once the application is refactored to fit inside an umbrella structure, and a new Phoenix app is added, the developer is immediately faced with the issue that the same port cannot be used for both services. A way of routing traffic is needed from one HTTP(S) port to the different Phoenix applications.
+
+# Solutions
+There are several solutions, each with different tradeoffs.
+
+Here are some that I came up with:
+
+1. Setup a reverse proxy with a popular HTTP daemon like NGINX or Apache2 and use different ports for each app.
+1. Set up a load balancer to forward to different ports for each app.
+1. Separate the applications into different repos, deployments and releases (this approach is suitable for large teams of 100+ developers).
+1. Use an internal Phoenix reverse proxy.
+
+Let's look at each option in more detail.
+
+## Load Balancer or Reverse Proxy
+Both these options require DevOps intervention or modifications to DevOps code or configs. Not everyone knows how to configure a load balancer or NGINX. Keeping the overhead of creating a new Phoenix application low is critical; otherwise, developers will avoid it and instead build up a monolith. There is also an issue with performance depending on the application because additional IO needs to be performed to proxy the requests.
+
+## Separate Releases or Separate Repos
+If a company has reached a level of scale where it can afford the overhead of deploying and maintaining multiple services, DevOps will need to be involved. This opens up the possibility of doing everything at the DevOps layer.
+
+Eventually, this level of separation is what we should be striving for as an end goal when we divide up our applications in our umbrella. But it doesn't mean we need to do everything in the DevOps layer because individual services still, in many cases, need an umbrella structure involving more than one application; therefore, internal proxying is not ruled out for services that require that level of complexity. So a hybrid approach of separate repos and releases where some are umbrellas with internal proxying is likely. 
+
+## Internal Reverse Proxy
+There are currently three reverse proxy strategies that exist in the ecosystem:
+
+1. [Roll your own](https://blog.appsignal.com/2019/04/16/elixir-alchemy-routing-phoenix-umbrella-apps.html)
+1. [MasterProxy](https://github.com/jesseshieh/master_proxy)
+1. [PhoenixReverseProxy](https://github.com/loopsocial/phoenix_reverse_proxy)
```

Comment:
link is broken

Reply:
Good catch. It's expected since we have to go through legal before this is made public.